spec/consensus/light-client/detection.md

```diff
@@ -1,3 +1,190 @@
-# Detection
+# Failure Detection
 
-TODO
+If there are more than 1/3 (or more) faulty validators, safety may be violated.
+This document describes how the light client can detect such violations (after
+the fact) and the next steps.
+
+## 1. Detection
+
+The bare minimum requirement is to be connected to at least one honest full
+node. This is sufficient so long as there's no fork on the main chain.
+
+If there is a fork on the main chain, it means that two full nodes have decided
+on two different headers for the same height. Then the requirement would at
+least require that the lite client is connected to one honest full node on
+each branch of the fork.
+
+_Remark_: +1/3 of malicious validators can create an unlimited number of forks.
+
+In practice, this means connecting to 1 or more geographically distributed full
+nodes (called **witnesses**), which belong to different companies. Note this
+number does not include the primary full node (called primary), which is used
+for obtaining new headers.
+
+_Remark_: we can't guarantee all witnesses won't follow the same branch of a
+fork. to guarantee that, we'll need the light client to be connected to +1/3 of
+nodes, which is impossible (_the structure of the network where validators are
+hidden behind sentries makes this impossible_).
+
+Full nodes are much more connected. And, if we assume a correct full node is
+going to halt in case of a fork on the main chain, then **we only need to
+notice that fact**.
+
+After the light client verifies a new header (`H`) it received from primary, it
+should cross-check `H` with the headers from all witnesses. Cross-checking
+means comparing hashes of the headers. If any two hashes (or more) diverge,
+there's a fork (on the main chain OR phantom fork targeting this light client).
+
+The light client will then need to validate the header it got from a witness
+(`H1`) and verify the signers account for +1/3 of the voting power.
```

Comment:
shall we add " and verify the signers account for +1/3 of the voting power of the validator set belonging to a header from the primary at the height of the root of the fork"? Is it important that we find the root of the fork in which we compare the validator set overlap?

Reply:
I'm thinking this may be the best way to go where we don't have to send traces along w/ SignedHeader's

Reply:
Okay, I'm thinking we have a reverse bisection function where we take both the secondary and primary and using bisection try to find the last header that they had in common and then submit the headers that come directly thereafter, in which there should be at least 1/3 overlap between both headers, as the conflicting headers - this way no matter where in the chain the light client recognises the fork all light clients will submit the same type of evidence - which happens at the root